executor/infoschema_reader.go

```diff
@@ -153,3 +162,164 @@ func dataForViews(ctx sessionctx.Context, schemas []*model.DBInfo) ([][]types.Da
 	}
 	return rows, nil
 }
+
+//slowQueryRetriever is used to read slow log data.
+type diskUsageRetriever struct {
+	dummyCloser
+	table         *model.TableInfo
+	outputCols    []*model.ColumnInfo
+	retrieved     bool
+	initialized   bool
+	extractor     *plannercore.DiskUsageExtractor
+	initialTables []*initialTable
+	curTable      int
+}
+
+func (e *diskUsageRetriever) retrieve(ctx context.Context, sctx sessionctx.Context) ([][]types.Datum, error) {
+	if e.retrieved {
+		return nil, nil
+	}
+	if !e.initialized {
+		err := e.initialize(sctx)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if len(e.initialTables) == 0 || e.curTable >= len(e.initialTables) {
+		e.retrieved = true
+		return nil, nil
+	}
+
+	rows, err := e.dataForDiskUsage(sctx)
+	if err != nil {
+		return nil, err
+	}
+	if len(e.outputCols) == len(e.table.Columns) {
+		return rows, nil
+	}
+	retRows := make([][]types.Datum, len(rows))
+	for i, fullRow := range rows {
+		row := make([]types.Datum, len(e.outputCols))
+		for j, col := range e.outputCols {
+			row[j] = fullRow[col.Offset]
+		}
+		retRows[i] = row
+	}
+	return retRows, nil
+}
+
+type initialTable struct {
+	db string
+	tb *model.TableInfo
+}
+
+func (e *diskUsageRetriever) initialize(sctx sessionctx.Context) error {
+	is := infoschema.GetInfoSchema(sctx)
+	dbs := is.AllSchemas()
+	sort.Sort(infoschema.SchemasSorter(dbs))
+	schemas := e.extractor.TableSchema
+	tables := e.extractor.TableName
+	//var initialTables []initialTable
+	if len(schemas) != 0 {
+		for _, schema := range dbs {
+			if schemas.Exist(schema.Name.L) {
```

Comment:
Is `schemas` guaranteed to be in lowercase?

Reply:
Yes, all `schema `in `schemas `have been turn into lowercase in `TableStorageStatsExtractor.Extract()`.